pkg/services/ngalert/notifier/channels/dingding.go

```diff
@@ -37,8 +32,8 @@ func NewDingDingNotifier(model *NotificationChannelConfig, ns notifications.Webh
 			DisableResolveMessage: model.DisableResolveMessage,
 			Settings:              model.Settings,
 		}),
-		MsgType: msgType,
-		URL:     url,
+		MsgType: model.Settings.Get("msgType").MustString(defaultDingdingMsgType),
+		URL:     model.Settings.Get("url").MustString(),
```

Comment:
I'm a bit concerned that the validation can be broken since we're parsing the JSON directly in two separate areas of code.

Say a developer wants to add a new config option to a channel. They'd navigate here to the channel file, see this existing code, and pull their new option out with `Get`. The notifier would appear to work, but they forgot to go modify the validation file! Nothing in the type system ever hinted to them that this file existed, and now we have a bug because users can write invalid config to the database - the validator isn't properly checking this new option.

What if we had the validation code produce structs instead, and had the notifier accept a pre-validated struct? Then, everything would be strongly typed. There'd only be one place for the parsing code to live, and devs would be forced to extend the validator if they want to change the settings. I think this will reduce the chance of bugs going forward, since we're using the type system to enforce validity.

Reply:
Struct embedding would work really nicely here too. Rather than having `DingDingNotifier` accept an un-validated NotificationChannelConfig, it could extend the global config with its own options. The validator would then produce this combined struct.
```
type DingDingConfig struct {
	NotificationChannelConfig
	MsgType string
	Url     string
	Message string
}
```

```
func NewDingDingNotifier(model *DingDingConfig, ...
```

Reply:
The validator for `dingding` would then return a `DIngDingConfig`. We'd keep the pattern where before writing to the DB, we run the validator and check for errors. Then, when constructing the `DingDingNotifier` we would then run the validator, get this config struct, and build the notifier.

The benefits are:
1. Extremely clear model. You cannot produce an instance of `DingDingConfig` without having run validation code, so any logic that takes a `DingDingConfig` is guaranteed to have run validation.
2. Extracts even more code from the channel constructors and unifies it. Optional parameters are also validated and substituted earlier on.
3. Separate validation methods for each channel type avoid the OCP issues in the other comment. It's now very easy to add new options or channels without having to touch much code.
4. If a dev wants to add a new option, they ahve to extend `DingDingConfig`, therefore extending the validator. The type system enforces that the extracted validation code stays up to date.

Reply:
That's basically what upstream is doing, I think I'll copy their approach for us here.

Reply:
That being said, it also does not force you to think about validation when you add a method 🤷 But I think it's a first step in the right direction.